torchao/prototype/low_bit_optim/adam.py

```diff
@@ -48,10 +49,17 @@ def step(self, closure=None):
                 if grad.is_sparse:
                     raise RuntimeError("Sparse gradient is not supported")
 
+                # unwrap DTensor
```

Comment:
cc: @wanchaol 

For context: What is the advantage of _not_ unwrapping the `DTensor`? The optimizer states ideally can be `DTensor`s too, in which case they have their sharding info present on the tensor, making optimizer-state checkpointing simplified.

However, low-bit optimizers may present challenges to this, which we should understand (from the `DTensor` side). It would be cool to understand the pain points and blockers (whether around functionality/UX in eager mode, compile support, etc.).

Reply:
@awgu Thanks for bringing this up! 

@gau-nernst  wondering if there's any challenges to not unwrapping DTensors? as Andrew mentioned, optimizer states should be DTensors so that it should be save/load correctly when performing distributed training. We have implemented the sharding strategies for Adam in core so if low-bit optimizers are similar to the Adam implementation in core, we just need to append the operators to the corresponding list

Reply:
I have tried using DTensor directly (that contains low-bit tensor subclass as its `local_tensor`) but faced the following error

> [rank0]:E0708 20:40:01.356000 128135323341888 torch/testing/_internal/common_distributed.py:664] AssertionError: s10 (could be from ["L['grad']._base._local_tensor.size()[1]"]) not in {s2: ["L['exp_avg']._local_tensor.codes.size()[0]", "L['exp_avg']._local_tensor.codes.size()[0]", "L['exp_avg']._local_tensor.codes.size()[0]", "L['exp_avg']._local_tensor.codes.size()[0]", "L['exp_avg']._local_tensor.codes.size()[0]"], s3: ["L['exp_avg']._local_tensor.scale.size()[0]", "L['exp_avg']._local_tensor.scale.size()[0]", "L['exp_avg']._local_tensor.scale.size()[0]", "L['exp_avg']._local_tensor.scale.size()[0]", "L['exp_avg']._local_tensor.scale.size()[0]"], s4: ["L['exp_avg']._local_tensor.qmap.size()[0]", "L['exp_avg']._local_tensor.qmap.size()[0]", "L['exp_avg']._local_tensor.qmap.size()[0]", "L['exp_avg']._local_tensor.qmap.size()[0]", "L['exp_avg']._local_tensor.qmap.size()[0]", "L['exp_avg_sq']._local_tensor.qmap.size()[0]", "L['exp_avg_sq']._local_tensor.qmap.size()[0]", "L['exp_avg_sq']._local_tensor.qmap.size()[0]", "L['exp_avg_sq']._local_tensor.qmap.size()[0]", "L['exp_avg_sq']._local_tensor.qmap.size()[0]"], s13: ["L['grad']._local_tensor.size()[0]", "L['grad']._local_tensor.size()[0]"], s11: ["L['grad']._local_tensor.storage_offset()", "L['grad']._local_tensor.storage_offset()"], s16: ["L['exp_avg_sq']._local_tensor.codes.size()[0]", "L['exp_avg_sq']._local_tensor.codes.size()[0]", "L['exp_avg_sq']._local_tensor.codes.size()[0]", "L['exp_avg_sq']._local_tensor.codes.size()[0]", "L['exp_avg_sq']._local_tensor.codes.size()[0]"], s17: ["L['exp_avg_sq']._local_tensor.scale.size()[0]", "L['exp_avg_sq']._local_tensor.scale.size()[0]", "L['exp_avg_sq']._local_tensor.scale.size()[0]", "L['exp_avg_sq']._local_tensor.scale.size()[0]", "L['exp_avg_sq']._local_tensor.scale.size()[0]"], s25: ["L['p']._local_tensor.size()[0]", "L['p']._local_tensor.size()[0]"]}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665

I'm guessing it's due to me using `.view(-1)` on the DTensor. If I remove torch.compile, the test passes again. Also, I'm not very clear if using `.view(-1)` is a proper way? How does DTensor handle `.view(-1)`? Does it keep the same local tensor, or it gathers the tensor and reshard? We probably don't want the latter.

Adding on about `.view(-1)`. I added this since torch.compile will recompile and eventually hit max cache size limit when facing different inputs, even though I tried to set `dynamic=True`. Perhaps we should try to fix this instead? (i.e. so that we don't need to use `.view(-1)` anymore, seems like the source of many problems).

I will make another branch using DTensor as optim state, so you can inspect and debug the error.

Reply:
Another option is init optim state as DTensor, but when call the Adam step, unwarp everything to local tensor. Kinda a workaround to get distributed saving/loading, while keeping torch.compile happy.

Reply:
sounds good, I think on a `Shard(0)` placement DTensor (where FSDP2 params look like), the `view(-1)` shouldn't do any communication/reshard and it should return a `Shard(0)` DTensor directly, so on eager side it should behave exactly what we want, otherwise it's a bug in DTensor.

I'm not sure why we hit recompliation when turning on torch.compile, it would be great if you could create a branch for repro, and we can take a look why compile not working expected.

cc @bdhirsh

Reply:
About the recompilation for torch.compile. To clarify, this is an issue for normal Tensor also, not only DTensor.

Reply:
@gau-nernst if I pull this PR locally, do you have a command that I can repro the compile error with? (both the error when not unwrapping DTensor here, and the recompiles with the low bit optimizer by itself).

I think it would be useful to understand what's going on here and why the dynamic shapes are showing up. The trace above isn't very helpful because we don't have a full stacktrace

Reply:
@bdhirsh You can refer to #490 for more details. To run the test, it's just `pytest "test/prototype/test_low_bit_optim.py::TestFSDP2::test_fsdp2"`.

Reply:
I looked at the error a little more.

For a bit more context, there are still two areas that we need to fix for to get proper enablement for "DTensor + dynamic shapes + compile":

(1) audit DTensor some more to find places where we are implicitly specializing dynamic shapes
(2) fix an existing gap in the "tensor subclass + compile + dynamic shape" surface area: today, things break if your compiled region returns a tensor subclass that has a dynamic shapes (see [issue](https://github.com/pytorch/pytorch/issues/124619))

On (1), I did a bit of auditing [here](https://www.internalfb.com/phabricator/paste/view/1475144406).

When I run the repro above, I actually don't get the same error that @gau-nernst hit, but I do get an error during tracing:
```
[rank2]:     return get_proxy_slot(e, tracer)()
[rank2]:   File "/home/hirsheybar/local/b/pytorch/torch/fx/experimental/proxy_tensor.py", line 137, in get_proxy_slot
[rank2]:     raise RuntimeError(f"{obj} is not tracked with proxy for {tracer}")
[rank2]: torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
[rank2]: RuntimeError: s0 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f94dff60c40>
```

Guilherme has been working on a fix for that issue here: https://github.com/pytorch/pytorch/pull/125941. Once it's landed, we should probably give this another try.

That raises the question: @gau-nernst how important is it to avoid recompiling on shape changes in your use case? When I run your repro with `torch.compile(..., dynamic=False)`, it runs successfully for me. I'm hoping that it will be useful if we can separate "low-bit optimizer + FSDP2 + compile working properly with static shapes", and fix the latent dynamic shape issues in the near future.

If that sounds acceptable, then one option is to land the version of this PR that does not unwrap DTensor, and then followup on getting dynamic shapes to work (cc @wanchaol since this seems like a candidate use case for dynamic shapes + DTensor being more important)

Reply:
Whether compile cache size limit is hit depends on the model. To be more specific, it is the variety of param shapes. There are 2 main sources of recompilation for my `single_param_adam()` functions:
1. Input shape changes (if dynamic=True, then it is when ndim - or tensor rank - changes)
2. Class object - we need to compile one version for normal tensor optim state, and another version for quantized tensor subclass optim state. This is because not all params will be qualified for quantization (e.g. too small, size is not divisible by block size, or user might want to exclude certain params like embedding layer)

Just 2 sources above can result in many combinations. For a ViT model from timm that I tested [here](https://github.com/pytorch/ao/blob/main/benchmarks/benchmark_low_bit_adam.py), whose params have a lot of different shapes, not doing `dynamic=True` and `.view(-1)` will hit compile cache size limit easily. Maybe for LLMs, params don't have such a variety of shapes, but it's kinda not very nice if we don't handle the case when there are a variety of param shapes.

A simple solution (without doing `.view(-1)`) is to just increase cache size limit. Not sure if it is encouraged?

Reply:
> A simple solution .... is to just increase cache size limit. Not sure if it is encouraged?

It probably depends on your use case. One failure-mode when increasing the cache limit is if e.g. your inputs have an unbounded amount of dynamism and you end up recompiling on every new input that comes in. But if you know that you have a fixed number of parameters with distinct shapes (e.g. ~50) and you know that you will only need to recompile once per param shape, bumping the cache size limit to what you need is probably ok

> There are 2 main sources of recompilation for my single_param_adam()

Mostly a curiosity question: is there anything stopping you from compiling the entire step function? Or maybe a subset of it, that performs `single_param_adam()` on every parameter? That way you get one giant graph with all of the params (allowing inductor to potentially fuse more things), instead of many small graphs.

Regardless, we want to fix the dynamic shape issues and this is probably a good reason to keep pushing on https://github.com/pytorch/pytorch/issues/124619

Reply:
> It probably depends on your use case. One failure-mode when increasing the cache limit is if e.g. your inputs have an unbounded amount of dynamism and you end up recompiling on every new input that comes in. But if you know that you have a fixed number of parameters with distinct shapes (e.g. ~50) and you know that you will only need to recompile once per param shape, bumping the cache size limit to what you need is probably ok

That's why I'm leaning towards the unwarping DTensor "hack", since `dynamice=True` + `.view(-1)` already work well (at least I think so) for non-FSDP setup, so it's natural to "hack" it to support DTensor.

> is there anything stopping you from compiling the entire step function?

I thought of it but haven't tried it myself. I can give it a try.

So in the end, how should we decide which approach to close this PR? Maybe I will try the horizontal fusion that you mention first to see if it can solve all of the problems.

Reply:
Actually, re-reading @awgu's comment that was not answered

> What is the advantage of **not** unwrapping the DTensor?

In other words, what are the problems that unwrapping DTensor might be discouraged? Quantized optim state can be wrapped by DTensor and then unwrapped before adam step too.

Reply:
> In other words, what are the problems that unwrapping DTensor might be discouraged? Quantized optim state can be wrapped by DTensor and then unwrapped before adam step too.

Ah, you are saying you can keep the optimizer states as `DTensor`s but just unwrap them for the actual `step()`? If so, I think that works. We would only not prefer it because it requires the optimizer code to be aware of `DTensor`. (Ideally, like in the native torch optimizers, we do not need to change the code to make it work with `DTensor`, but this is just the ideal state.)